### PR TITLE
Add Schwinger process

### PR DESCRIPTION
**Pull request by NeilZaim**
**Overview**

This PR implements the Schwinger pair creation process in WarpX. If the code is compiled with QED and the user activates the Schwinger process in the input file, electron-positron pairs can be created in the function `MultiParticleContainer::doQEDSchwinger` which is called at every time step in `Evolve`.
`MultiParticleContainer::doQEDSchwinger` in turn calls the function `filterCreateTransformFromFAB` which has been written for this purpose (heavily based on `filterCopyTransform`). 
 `filterCreateTransformFromFAB` proceeds in three steps. In the filter phase, we loop on every cell and calculate the number of physical pairs created within the time step dt as a function of the electromagnetic field at the given cell position. This probabilistic calculation is done via a wrapper that calls the PICSAR library (for a ref regarding the analytical formula used in the pair production rate see e.g. [here](https://journals.aps.org/prl/abstract/10.1103/PhysRevLett.104.220404)). In the create phase, the particles are created at the desired positions, currently at the cell nodes. In the transform phase, we assign a weight to the particles depending on the number of physical pairs created. At most one macroparticle is created per cell per timestep per species, with a weight corresponding to the total number of physical pairs created.

So far the Schwinger module requires using `warpx.do_nodal=1` (actually it also works with  `warpx.do_nodal=0` provided that we use the momentum conserving algorithm, see edit) and is not compatible with either mesh refinement, RZ coordinates or single precision (in this case because one of the constants used in the pair production rate calculation overflows in PICSAR). The next step will probably be to make the module work without `do_nodal` (in a separate PR). For that we will have to interpolate all the field components at the same position on the grid before computing the pair production rate.

**Usage**

Schwinger process is activated by adding `warpx.do_qed_schwinger = 1` in the input file. In this case, the user must also specify the product electron and positron species in `qed_schwinger.ele_product_species` and `qed_schwinger.pos_product_species` respectively. These species must be defined in `particles.species_names` and must actually correspond to an electron and a positron species.

If the simulation is run in 2D, the user must also specify a transverse size used in the pair production rate calculation. This is done through the parameter `qed_schwinger.y_size`.

An additional parameter is `qed_schwinger.threshold_poisson_gaussian` which is optional and equal to 100 by default. If the expected number of physical pairs created in a cell at a given timestep is smaller than this threshold, a Poisson distribution is used to draw the actual number of physical pairs created. Otherwise a Gaussian distribution is used.

**Testing**

A test of this module has been added to the suite. It consists of 4 light 3D simulations with a total size of 8\*8\*8 cells, a total of 1 timestep and an external uniform EM field in vacuum. In the first test, the field is "low" so that no particles are created. In the second test, the field is extremely high and a really high number of (physical) particles are created (a Gaussian distribution is used in this case). In the third test, the field is intermediate and an average of 1 pair per cell is created (with a Poisson distribution). In the fourth test, the EM field is extremely high but the E and B field are perpendicular with a comparable amplitude so that the number of created pairs is rather low (with a Gaussian distribution). In the test, we only verify that the total generated charge corresponds to that expected (within a 5 sigma bound) but I have also checked visually that the distributions correspond to those expected.

These tests work both on CPU and GPU. I have also tried a few other things (split the domain into several boxes, use more than 1 timestep, activate the field only in part of the domain) and it all seems to work properly.

PS: I have left a few questions in the comments of the code regarding specific points. I guess I'll remove them before merging but feel free to leave a reply during the review. :) 

**Edit** : as it turns out the code already works without `warpx.do_nodal=1` provided that we use the momentum conserving algorithm for field gathering (we actually require that the auxiliary fields be on a nodal grid). I've changed the asserts so that it works in this case and I've also changed the test so that it is compiled without PSATD (I guess it should save one compilation during testing).

### CODE REVIEW

File: Source/Particles/MultiParticleContainer.cpp [hunk 1]
```diff
@@ -991,6 +1011,112 @@ MultiParticleContainer::BreitWheelerGenerateTable ()
...
+    const MultiFab & Ez = warpx.getEfield(level_0,2);
+    const MultiFab & Bx = warpx.getBfield(level_0,0);
+    const MultiFab & By = warpx.getBfield(level_0,1);
+    const MultiFab & Bz = warpx.getBfield(level_0,2);
+
+        // get cell size multiplied by time step
+    Geometry const & geom = warpx.Geom(level_0);
+    auto domain_box = geom.Domain();
+#if (AMREX_SPACEDIM == 2)
+    auto dVdt = geom.CellSize(0) * geom.CellSize(1) * warpx.getdt(0); // getdt(level_0) ??
```
Comment (lucafedeli88) [inline Source/Particles/MultiParticleContainer.cpp]: I am still not completely sure about what to do for dVdt in 2D...

File: Source/Particles/MultiParticleContainer.cpp [hunk 2]
```diff
@@ -991,6 +1011,112 @@ MultiParticleContainer::BreitWheelerGenerateTable ()
...
+        auto arrNumPairCreation = NumPairCreation.array();
+
+        amrex::ParallelFor(box,  [=] AMREX_GPU_DEVICE (int i, int j, int k){
+                arrNumPairCreation(i,j,k) = getSchwingerProductionNumber( dVdt,
+                    arrEx(i,j,k),arrEy(i,j,k),arrEz(i,j,k),
+                    arrBx(i,j,k),arrBy(i,j,k),arrBz(i,j,k));
+                amrex::Print(arrNumPairCreation(i,j,k));
+                }); */
+
+        auto& dst_ele_tile = pc_product_ele->ParticlesAt(level_0, mfi);
```
Comment (lucafedeli88) [inline Source/Particles/MultiParticleContainer.cpp]: Are we sure that adding at level 0 is OK for particles?

File: Source/Particles/MultiParticleContainer.cpp [hunk 3]
```diff
@@ -991,6 +1011,112 @@ MultiParticleContainer::BreitWheelerGenerateTable ()
...
+        auto& dst_pos_tile = pc_product_pos->ParticlesAt(level_0, mfi);
+
+        int dst1_index_dummy=0;
+        int dst2_index_dummy=0;
+        auto filter_dummy= [](auto, auto, auto, auto) {return 0;};
+        auto create_dummy= [](auto, auto, auto, auto, auto) {return 0;};
+        auto transform_dummy= [](auto, auto, auto, auto) {return 0;};
```
Comment (lucafedeli88) [inline Source/Particles/MultiParticleContainer.cpp]: I like this :laughing:

File: Source/Particles/ElementaryProcess/QEDInternals/SchwingerProcessWrapper.H [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#ifndef WARPX_schwinger_process_wrapper_h_
+#define WARPX_schwinger_process_wrapper_h_
+
+#include "QedWrapperCommons.H"
+
+//#define PXRMP_CORE_ONLY allows importing only the 'core functions' of the
+//Quantum Synchrotron engine of the QED PICSAR library.
```
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Source/Particles/ElementaryProcess/QEDInternals/SchwingerProcessWrapper.H]: Done :)

File: Source/Particles/MultiParticleContainer.cpp [hunk 5]
```diff
@@ -991,6 +1011,112 @@ MultiParticleContainer::BreitWheelerGenerateTable ()
     }
 }
 
+void
+MultiParticleContainer::InitSchwinger ()
```
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Source/Particles/MultiParticleContainer.cpp]: Yes that's a good idea, I've done that.

File: Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+                                    CreateFunc&& create,
+                                    TransFunc && transform) noexcept
+{
+    using namespace amrex;
+
+    // FAB representing the number of particles to create
+    FArrayBox NumPartCreation(box, 1);
+    auto arrNumPartCreation = NumPartCreation.array();
+
+    const Dim3 lo = lbound(box);
```
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H]: Thanks again, I've done that.

File: Source/Particles/MultiParticleContainer.cpp (hunk 2 above)
Comment (NeilZaim) [reply Source/Particles/MultiParticleContainer.cpp]: No :) (I'm not sure if the particles will be put at the correct level the next time Redistribute() is called)
Let's assume for now that this module only works without mesh refinement (I've added an assert for that), we can modify this later on.

**Comment by ax3l** [comment]
@NeilZaim can you please add a pull-request description? :)

**Comment by MaxThevenet** [comment]
It's still [WIP], maybe it's hard to right a description already :).

**Comment by NeilZaim** [comment]
I've just added a quick description. But as Maxence said, this PR is not finished yet; I will write a more complete description whenever it is ready for reviewing.

File: Source/Particles/ElementaryProcess/QEDSchwingerProcess.H [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+{
+
+    const amrex::Real m_y_size;
+
+    template <typename DstData1, typename DstData2>
+    AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+    void operator() (DstData1& dst1, DstData2& dst2, int i_dst1, int i_dst2,
+                               int N, amrex::Real total_weight) const noexcept
+    {
+        const auto m_weight_index = ParticleStringNames::to_index.find("w")->second;
```
Comment (lucafedeli88) [inline Source/Particles/ElementaryProcess/QEDSchwingerProcess.H]: I think that I would just call it `weight_index`, since it is not a member variable of the `struct SchwingerTransformFunc`

File: Source/Particles/ElementaryProcess/Make.package [hunk 8]
```diff
@@ -1,7 +1,10 @@
 ifeq ($(QED),TRUE)
+<<<<<<< HEAD
```
Comment (ax3l) [inline Source/Particles/ElementaryProcess/Make.package]: careful: you committed an unresolved merge conflict :)

Comment (NeilZaim) [reply Source/Particles/ElementaryProcess/Make.package]: Thanks. :) Though they should be resolved in the last commit no?

Comment (ax3l) [reply Source/Particles/ElementaryProcess/Make.package]: Gone now, maybe I commented on an outdated commit.

Comment (ax3l) [reply Source/Particles/ElementaryProcess/Make.package]: Gone now, maybe I commented on an outdated commit.

File: Examples/Modules/qed/schwinger/analysis.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+import glob
+import yt
+import numpy as np
+
+# define some parameters
+
+c = 299792458.
+m_e = 9.10938356e-31
+e = 1.6021766208e-19
+hbar = 1.054571800e-34
```
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Examples/Modules/qed/schwinger/analysis.py:20]: The problem is that the second test is extremely sensitive and fails if I use scipy constants (which are slightly different from the ones in picsar).

This might not be ideal, especially if we someday change the constants in picsar (don't know why we would do that) or if we want to make Schwinger work in single precision (which may result in the second test failing). If you think it's better I can reduce the sensitivity of the second test by reducing the number of pairs created per cell.

File: Examples/Modules/qed/schwinger/analysis.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+Bz_test3 = 0.
+
+Ex_test4 = 0.
+Ey_test4 = 0.
+Ez_test4 = 2.5e+20
+Bx_test4 = 0.
+By_test4 = 833910154604.3563
+Bz_test4 = 0.
+
+def calculate_rate(Ex,Ey,Ez,Bx,By,Bz):
```
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Examples/Modules/qed/schwinger/analysis.py:55]: Done that, thanks.

File: Source/Particles/ElementaryProcess/QEDSchwingerProcess.H [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+ * calculate the number of pairs created during a given timestep at a given cell.
+ */
+struct SchwingerFilterFunc
+{
+    int m_threshold_poisson_gaussian;
+    const amrex::Real m_dVdt;
+
+    /** Get the number of created pairs in a given cell at a given timestep.
+     *
+     * @param[in] src_FABs Array of 6 Array4 that contain the EM field in the tile.
```
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Source/Particles/ElementaryProcess/QEDSchwingerProcess.H:26]: Done that, thanks.

File: Source/Particles/ElementaryProcess/QEDSchwingerProcess.H [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+ * the Schwinger process.
+ */
+struct SchwingerTransformFunc
+{
+    const amrex::Real m_y_size;
+    const int m_weight_index;
+
+    /** Assign a weight to particles created via the Schwinger process.
+     *
+     * @param[in,out] dst1 Target electron species.
```
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Source/Particles/ElementaryProcess/QEDSchwingerProcess.H:64]: Done that, thanks.

File: Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+ * \param[in] box the box where the particles are created
+ * \param[in] src_FAB A FAB defined on box that is used in the transform function
+ * \param[in] mask pointer to the mask - 1 means create, 0 means don't create
+ * \param[in] dst1_index the location at which to starting writing the result to dst1
+ * \param[in] dst2_index the location at which to starting writing the result to dst2
+ * \param[in] create1 callable that defines what will be done for the create step for dst1.
+ * \param[in] create2 callable that defines what will be done for the create step for dst2.
+ * \param[in] transform callable that defines the transformation to apply on dst1 and dst2.
+ *
+ * \return num_added the number of particles that were written to dst 1 and dst 2.
```
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H]: Done that, thanks.

File: Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+ * \tparam Index the index type, e.g. unsigned int
+ * \tparam CreateFunc1 the create function type for dst1
+ * \tparam CreateFunc2 the create function type for dst2
+ * \tparam TransFunc the transform function type
+ *
+ * \param[in,out] dst1 the first destination tile
+ * \param[in,out] dst2 the second destination tile
+ * \param[in] box the box where the particles are created
+ * \param[in] src_FAB A FAB defined on box that is used in the transform function
+ * \param[in] mask pointer to the mask - 1 means create, 0 means don't create
```
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H]: It is `1`, thanks for pointing out the vagueness. I've changed it (and also in `FilterCopyTransform.H` from which I have copied this line).

File: Source/Particles/ElementaryProcess/QEDSchwingerProcess.H (hunk 12 above)
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Source/Particles/ElementaryProcess/QEDSchwingerProcess.H:64]: Done that, thanks.

File: Regression/WarpX-tests.ini [hunk 15]
```diff
@@ -894,6 +894,23 @@ compareParticles = 0
...
+buildDir = .
+inputFile = Examples/Modules/qed/schwinger/analysis.py
+aux1File = Examples/Modules/qed/schwinger/inputs_3d_schwinger
+customRunCmd = ./analysis.py
+dim = 3
+addToCompileString = USE_PSATD=TRUE QED=TRUE
+restartTest = 0
+useMPI = 1
+useOMP = 1
+numthreads = 2
```
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Regression/WarpX-tests.ini]: You're right, I've changed that. Additionally, there should be a single tile in these tests so it probably doesn't make much sense to use several threads.

File: Examples/Modules/qed/schwinger/analysis.py [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+        return  e**2*E_S**2/4./np.pi**2/c/hbar**2*epsilon**2/np.pi*np.exp(-np.pi/epsilon)
+
+
+def do_analysis(Ex,Ey,Ez,Bx,By,Bz):
+
+    expected_total_physical_pairs_created = dV*dt*calculate_rate(Ex,Ey,Ez,Bx,By,Bz)
+
+    if expected_total_physical_pairs_created < 0.01:
+        assert(not os.path.isdir(filename+"/ele_schwinger/Level_0/"))
+        ## Assert whether pairs are created or not. Is there a cleaner way to do that?
```
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Examples/Modules/qed/schwinger/analysis.py]: Done that, thanks.

File: Examples/Modules/qed/schwinger/analysis.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Examples/Modules/qed/schwinger/analysis.py:55]: Done that, thanks.

File: Docs/source/running_cpp/parameters.rst [hunk 17]
```diff
@@ -1502,6 +1502,35 @@ Lookup tables store pre-computed values for functions used by the QED modules.
...
+    If Schwinger process is activated, a positron product species must be specified
+    (the name of an existing positron species must be provided).
+
+* ``qed_schwinger.y_size`` (`float`; in meters)
+    If Schwinger process is activated with ``DIM=2D``, a transverse size must be specified.
+    It is used to convert the pair production rate per unit volume into an actual number of created particles.
+    This value should correspond to the typical transverse extent for which the EM field has a very high value
+    (e.g. the beam waist for a focused laser beam).
+
+* ``qed_schwinger.threshold_poisson_gaussian`` (`integer`) optional (default `100`)
```
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Docs/source/running_cpp/parameters.rst]: You're right. I've tested and the Schwinger function was indeed 2 to 3 times slower when the expected number of pairs per cell was 99 compared to when it was 101. I've changed the default value to 25, which should be high enough to have rare unphysical pair creations (with weight < 1) extremely rare (~ 5 sigma).

File: Examples/Modules/qed/schwinger/analysis.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Examples/Modules/qed/schwinger/analysis.py:20]: Ah! The discrepancy is due to the fact that constants in PICSAR come from CODATA 2014, while scipy (and WarpX) use newer data (CODATA 2018). 
I think that we could leave the hard-coded constants, but we should also consider to update the values of the fundamental constants in PICSAR!

File: Examples/Modules/qed/schwinger/analysis.py [hunk 18]
```diff
@@ -72,23 +74,22 @@ def calculate_rate(Ex,Ey,Ez,Bx,By,Bz):
...
 def do_analysis(Ex,Ey,Ez,Bx,By,Bz):
 
-    expected_total_physical_pairs_created = dV*dt*calculate_rate(Ex,Ey,Ez,Bx,By,Bz)
+    data_set = yt.load(filename)
 
+    expected_total_physical_pairs_created = dV*dt*calculate_rate(Ex,Ey,Ez,Bx,By,Bz)
     if expected_total_physical_pairs_created < 0.01:
-        assert(not os.path.isdir(filename+"/ele_schwinger/Level_0/"))
-        ## Assert whether pairs are created or not. Is there a cleaner way to do that?
+        assert(not data_set.particle_type_counts)
```
Comment (lucafedeli88) [inline Examples/Modules/qed/schwinger/analysis.py:81]: Uhm, `ds.particle_type_counts` should be a dictionary. I think that I would check the value of `data_set.particle_type_counts["ele_schwinger"]` and `data_set.particle_type_counts["pos_schwinger"]`

File: Source/Particles/ElementaryProcess/QEDSchwingerProcess.H [hunk 19]
```diff
@@ -65,9 +69,9 @@ struct SchwingerTransformFunc
      * @param[in] total_weight Number of physical pairs created in the
      *            considered cell.
      */
-    template <typename DstData1, typename DstData2>
+    template <typename DstTile>
     AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
-    void operator() (DstData1& dst1, DstData2& dst2, const int i_dst1,
+    void operator() (DstTile& dst1, DstTile& dst2, const int i_dst1,
```
Comment (lucafedeli88) [inline Source/Particles/ElementaryProcess/QEDSchwingerProcess.H:74]: Is there a particular reason for having `dst1` and `dst2` of the same type here?

Comment (NeilZaim) [reply Source/Particles/ElementaryProcess/QEDSchwingerProcess.H:74]: Not really, it was just to remain consistent with `FilterCreateTransformFromFAB` and `FilterCopyTransform` which both use the same template parameters for the type of dst1 and dst2. I can change that if you think it's better.

File: Examples/Modules/qed/schwinger/analysis.py (hunk 18 above)
Comment (MaxThevenet) [reply Examples/Modules/qed/schwinger/analysis.py:81]: any update on this discussion?

Comment (NeilZaim) [reply Examples/Modules/qed/schwinger/analysis.py:81]: Yes (sorry we discussed it on our side and didn't post an update on github).
It turns out that if there are no particles at all `particle_type_counts` returns an empty dict so this assert is correct. Using `particle_type_counts["ele_schwinger"]` would result in a KeyError.
(however if there were dumped particles of another species, we would have had to assert `particle_type_counts["ele_schwinger"] == 0`)

File: Source/Particles/MultiParticleContainer.cpp [hunk 20]
```diff
@@ -991,6 +1020,114 @@ MultiParticleContainer::BreitWheelerGenerateTable ()
...
+
+        const auto num_added = filterCreateTransformFromFAB<1>( dst_ele_tile,
+                              dst_pos_tile, box, array_EMFAB, np_ele_dst,
+                               np_pos_dst,Filter, CreateEle, CreatePos,
+                                Transform);
+
+        setNewParticleIDs(dst_ele_tile, np_ele_dst, num_added);
+        setNewParticleIDs(dst_pos_tile, np_pos_dst, num_added);
+
+        WARPX_PROFILE_VAR_STOP(blp_schwinger);
```
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Source/Particles/MultiParticleContainer.cpp:1127]: Well I defined the variable on purpose so that Schwinger does not pop up on the parser output if Schwinger process is turned off (I figured it might be slightly confusing for a user). Do you think I should remove it?

In any case you're right, the `STOP` is in the for loop. :-/ I'll fix that if we decide to keep the variable.

File: Source/Particles/ElementaryProcess/QEDInternals/SchwingerProcessWrapper.H [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+ * @param[in] PoissonToGaussianThreshold If the expected number of created
+ *            pairs is below this parameter, a Poisson distribution is used
+ *            to draw the number of created pairs. Otherwise a Gaussian
+ *            distribution is used.
+ */
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE amrex::Real
+getSchwingerProductionNumber (const amrex::Real dVdt,
+              const amrex::Real Ex, const amrex::Real Ey, const amrex::Real Ez,
+              const amrex::Real Bx, const amrex::Real By, const amrex::Real Bz,
+              const amrex::Real PoissonToGaussianThreshold = 100.)
```
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Source/Particles/ElementaryProcess/QEDInternals/SchwingerProcessWrapper.H:41]: I agree with both points, I'll commit your suggestion.

File: Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H [hunk 22]
```diff
@@ -0,0 +1,190 @@
...
+    const auto box_sizey = box.length(1);
+
+    auto & warpx = WarpX::GetInstance();
+    const int level_0 = 0;
+    Geometry const & geom = warpx.Geom(level_0);
+
+    Real xlo_global = geom.ProbLo(0);
+    Real ylo_global = geom.ProbLo(1);
+    Real dx         = geom.CellSize(0);
+    Real dy         = geom.CellSize(1);
```
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H:60]: Hmm no I probably got confused with the convention. I think I was writing this using the AMReX doc which I think doesn't have this convention. If you think it's better I can change this (that would add an extra `#if (AMREX_SPACEDIM == 2)` right ?) or at least add a comment specifying that we don't use the usual convention here, in case we want to change this in the future.

File: Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H [hunk 23]
```diff
@@ -0,0 +1,190 @@
...
+ */
+template <int N, typename DstTile, typename FABs, typename Index,
+          typename FilterFunc, typename CreateFunc1, typename CreateFunc2,
+           typename TransFunc>
+Index filterCreateTransformFromFAB (DstTile& dst1, DstTile& dst2, const amrex::Box box,
+                                const FABs& src_FABs, const Index dst1_index,
+                                const Index dst2_index, FilterFunc&& filter,
+                                CreateFunc1&& create1, CreateFunc2&& create2,
+                                TransFunc && transform) noexcept
+{
```
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H:155]: To be honest I'm not completely sure why we have the overload so I wouldn't know what to add as a comment for that (initially I wrote a single function but subsequently decided to keep the overload to remain consistent with `filterCopyTransform`). I guess it might be useful if we want to call the function using an externally obtained mask rather than a filter function.

I'll add a few comments in front of the for loops.

File: Examples/Modules/qed/schwinger/analysis.py (hunk 18 above)
Comment (MaxThevenet) [reply Examples/Modules/qed/schwinger/analysis.py:81]: This is just an analysis script, so we don't have to cover all corner cases, so it is fine to keep it this way. Otherwise, would this be a slightly more robust test?
```py
np_ele = data_set.particle_type_counts["ele_schwinger"] if \
    "ele_schwinger" in data_set.particle_type_counts.keys() else 0
np_pos = data_set.particle_type_counts["pos_schwinger"] if \
    "pos_schwinger" in data_set.particle_type_counts.keys() else 0
assert(np_ele == 0 and np_pos == 0)
```
Sorry if that's a stupid suggestion.

File: Source/Particles/MultiParticleContainer.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (MaxThevenet) [reply Source/Particles/MultiParticleContainer.cpp:1127]: If you put `WARPX_PROFILE` after the conditional return `if (!m_do_qed_schwinger) return;` (i.e., not like my suggestion) does the string still pop up in the Tiny Profiler when the Schwinger process is off?

Even if it does, I think it's nicer to use `WARPX_PROFILE("...")` as this is precisely meant for what we're doing here (profile a whole function). If the behavior is unexpected, we could consider fixing `WARPX_PROFILE`.

File: Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H (hunk 23 above)
[the remark this answers is not in the document]
Comment (MaxThevenet) [reply Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H:155]: @lucafedeli88 @atmyers could you help us understand the purpose of this overload?

File: Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H (hunk 22 above)
[the remark this answers is not in the document]
Comment (MaxThevenet) [reply Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H:60]: Either would be fine. And we know this convention (not applied everywhere, and inconsistent with AMReX) is a problem in WarpX, this is an active topic of discussion these days.

File: Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H (hunk 23 above)
[the remark this answers is not in the document]
Comment (lucafedeli88) [reply Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H:155]: I am not completely sure either . However, I agree with @NeilZaim : this implementation leaves the freedom to choose between providing a filter function or a pre-computed mask...

File: Examples/Modules/qed/schwinger/analysis.py (hunk 18 above)
Comment (NeilZaim) [reply Examples/Modules/qed/schwinger/analysis.py]: I've committed your suggestion. In the end it doesn't change much but it's more explicit and I guess it's also more robust (in case `yt` changes its behavior when no particles are present).

File: Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H (hunk 23 above)
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H:166]: I've added a few comments regarding the overloads (also in `FilterCopyTransform.H`) and I've also added a few comments regarding the for loops.

File: Source/Particles/MultiParticleContainer.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Source/Particles/MultiParticleContainer.cpp]: Ok thanks, I've committed your solution and removed the variable.

To answer your question: yes Schwinger did not appear in the Tiny Profiler when `WARPX_PROFILE` was after the `return` statement and the Schwinger process was turned off. Now (with the newest commit), Schwinger appears in the profiler even when the process is turned off.

File: Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H (hunk 22 above)
[the remark this answers is not in the document]
Comment (NeilZaim) [reply Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H]: Ok thanks, I've simply added a comment stating that we use x and y in 2D here.

File: Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H (hunk 23 above)
[the remark this answers is not in the document]
Comment (atmyers) [reply Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H:168]: For `filterCopyTransformParticles`, we made there be two versions of this function, in case someone already had a pre-computed mask and wanted to use that as the filter instead of a functor. The functor version just computes it's own mask and calls the other overload. That's what the ParalleFor is doing. 

I think it could potentially be useful to have both versions of these, but don't feel too strongly about it.

File: Source/Particles/MultiParticleContainer.H [hunk 24]
```diff
@@ -55,6 +55,9 @@ class MultiParticleContainer
 
 public:
 
+    using AllContainerType =
+        amrex::Vector<std::unique_ptr<WarpXParticleContainer>>;
```
Comment (MaxThevenet) [inline Source/Particles/MultiParticleContainer.H:59]: It seems that this is used only once. Is the goal to use it in more places? Otherwise, could we simply use `amrex::Vector<std::unique_ptr<WarpXParticleContainer>>`?

File: Source/Particles/MultiParticleContainer.cpp [hunk 25]
```diff
@@ -991,6 +1020,112 @@ MultiParticleContainer::BreitWheelerGenerateTable ()
...
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(warpx.maxLevel() == level_0,
+        "ERROR: Schwinger process not implemented with mesh refinement");
+
+#ifdef WARPX_DIM_RZ
+    amrex::Abort("Schwinger process not implemented in rz geometry");
+#endif
+
+#ifdef AMREX_USE_FLOAT
+    amrex::Abort("Schwinger process not implemented in single precision");
+#endif
```
Comment (MaxThevenet) [inline Source/Particles/MultiParticleContainer.cpp:1048]: <3

File: Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H [hunk 26]
```diff
@@ -0,0 +1,206 @@
...
+    using namespace amrex;
+
+    FArrayBox NumPartCreation(box, 1);
+    Elixir tmp_eli = NumPartCreation.elixir(); // Do we actually need this elixir?
+                            // The tests on GPU seem to be working well without it.
```
Comment (MaxThevenet) [inline Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H:173]: It is probably not required because you added an explicit `synchronize` in the other `filterCreateTransformFromFAB`. I would be in favor in keeping it though, with a comment
```suggestion
    // This may be unnecessary because of the Gpu::streamSynchronize() in
    // filterCreateTransformFromFAB called below, but let us keep it for safety.
    Elixir tmp_eli = NumPartCreation.elixir();
```

File: Source/Particles/MultiParticleContainer.H (hunk 24 above)
Comment (NeilZaim) [reply Source/Particles/MultiParticleContainer.H:59]: You're right, the initial goal was to use it multiple times but it turns out I didn't need to so I'll just remove it.

File: Regression/WarpX-tests.ini [hunk 27]
```diff
@@ -861,6 +861,23 @@ compareParticles = 0
...
 
+[qed_schwinger]
+buildDir = .
+inputFile = Examples/Modules/qed/schwinger/analysis.py
+aux1File = Examples/Modules/qed/schwinger/inputs_3d_schwinger
+customRunCmd = ./analysis.py
+dim = 3
+addToCompileString = QED=TRUE
+restartTest = 0
+useMPI = 1
```
Comment (MaxThevenet) [inline Regression/WarpX-tests.ini:872]:
```suggestion
useMPI = 1
numprocs = 1
```

**Comment by MaxThevenet** [comment]
I re-added the [WIP] tag so we don't merge this PR. After offline discussions with @NeilZaim, we will simplify the test to avoid running from a Python script, which is more complicated and causes issues, and will just write 4 separated tests.

File: Examples/Modules/qed/schwinger/analysis_schwinger.py [hunk 28]
```diff
@@ -0,0 +1,104 @@
...
+
+        ele_data = all_data["ele_schwinger",'particle_weight']
+        pos_data = all_data["pos_schwinger",'particle_weight']
+
+        std_total_physical_pairs_created = np.sqrt(expected_total_physical_pairs_created)
+
+        # This first assert only works if a single tile is used in the simulation
+        assert(np.array_equal(ele_data,pos_data))
+        # 5 sigma test that has an intrisic probability to fail of 1 over ~2 millions
+        assert(np.abs(np.sum(ele_data)-expected_total_physical_pairs_created)<5*std_total_physical_pairs_created)
```
Comment (MaxThevenet) [inline Examples/Modules/qed/schwinger/analysis_schwinger.py:102]: Could you print the errors before asserting? That way, even if the test fails on TravisCI we see by how much.

File: Examples/Modules/qed/schwinger/analysis_schwinger.py [hunk 29]
```diff
@@ -0,0 +1,104 @@
...
+    Bx_test = 1679288857.0516706
+    By_test = 525665014.1557486
+    Bz_test = 1836353079.9561853
+elif test_number == '3':
+    Ey_test = 1.0321239524474501e+17
+elif test_number == '4':
+    Ez_test = 2.5e+20
+    By_test = 833910154604.3563
+else:
+    assert(False)
```
Comment (MaxThevenet) [inline Examples/Modules/qed/schwinger/analysis_schwinger.py:56]: Could you add a quick description for each test? e.g.,
```py
if test_number == '1':
    // Low-field test, no particle should be created
```
(or whatever comment is relevant)
